### PR TITLE
Added pagination to `List` RPC calls

### LINKED ISSUE #635: gRPC `list` functions: Take into account max message length of received messages
**Describe the bug**
When generating so many assessment results (AR) that the list of a ARs in `ListAssessementResults()` is bigger than 4MB (4194304 Bytes), gRPC will throw an error. This error, of course, concerns all list calls.

**To Reproduce**
Collect/Generate a lot of evidences/ARs.

**Expected behavior**
Always return lists that are not exceeding 4MB in size. Either cut older ones or implement a mechanism to retrieve multiple lists.

**Additional context**
We could also increase the threshold for maximum message length but that would only delay the problem.

### PR DESCRIPTION
**Pull request by oxisto**
This PR adds pagination to all `List` RPC calls, even if they might not (yet) need it.

- [x] Implemented pagination for all RPC calls that directly return maps and values from memory
- [x] Implemented pagination for all RPC calls that return values from the `persistence.Storage` interface
- [x] Handling pagination on the client side, i.e., in the `cl` binary

Closes #635

### CODE REVIEW

**Comment by lebogg** [comment]
Closes #635

File: service/paginate.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+		token = nil
+	}
+
+	// Only needed, if more pages exist
+	if token != nil {
+		// Move the token "forward"
+		token.Start = end
+
+		// Encode next page token
+		nbt, err = token.Encode()
```
Comment (lebogg) [inline service/paginate.go:89]: npt?

File: cli/commands/cloud/cloud.go [hunk 2]
```diff
@@ -96,7 +98,14 @@ func NewListCloudServicesCommand() *cobra.Command {
 
 			client = orchestrator.NewOrchestratorClient(session)
 
-			res, err = client.ListCloudServices(context.Background(), &orchestrator.ListCloudServicesRequest{})
+			services, err = api.ListAllPaginated(&orchestrator.ListCloudServicesRequest{}, client.ListCloudServices, func(res *orchestrator.ListCloudServicesResponse) []*orchestrator.CloudService {
+				return res.Services
+			})
+
+			// Build a response with all services
+			res = &orchestrator.ListCloudServicesResponse{
```
[the remark this answers is not in the document]
Comment (oxisto) [reply cli/commands/cloud/cloud.go:106]: This is done on the client side, just to display the full list of services in the JSON output. The message size problem should be on the gRPC call.

File: api/evidence/evidence_store.go [hunk 3]
```diff
@@ -0,0 +1 @@
+package evidence
```
[the remark this answers is not in the document]
Comment (oxisto) [reply api/evidence/evidence_store.go:1]: Left over, can be deleted

File: persistence/persistence.go [hunk 4]
```diff
@@ -54,6 +54,10 @@ type Storage interface {
 	// List lists all records in database which meet the (optionally) given conditions
 	List(r interface{}, conds ...interface{}) error
 
+	// ListLimOff lists all records in database which meet the (optionally) given conditions with a certain limit after an offset.
+	// TODO(oxisto): find a proper name for this function
+	ListLimOff(r interface{}, offset int, limit int, conds ...interface{}) error
```
[the remark this answers is not in the document]
Comment (oxisto) [reply persistence/persistence.go:59]: How would we provide the default values?

File: cli/commands/cloud/cloud.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lebogg) [reply cli/commands/cloud/cloud.go:106]: I thought so as well. But when I do CLI commands they are gRPC calls. At least, they are represented in the Orchestrator log as gRPC calls and in addition, I got the error via CLI commands as well.

File: persistence/persistence.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lebogg) [reply persistence/persistence.go:59]: `List(&evidences, 0, 0, conds...)` would call (in gorm implementation) the list function and, e.g., `List(&evidences, 4, 3, conds...)`  the Offset function.

File: cli/commands/cloud/cloud.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (oxisto) [reply cli/commands/cloud/cloud.go:106]: Yes but the merging of the results is on the client side, so should be no problem. The message size only affects the transport and the transport is chunked / paged.

**Review by garuppel — APPROVED** [review]
:+1:

**Pull request merged**